### ISSUE TITLE
[ORC] Perf profiling integration not working in lli

### ISSUE THREAD

**Issue by tetzank** [llvm/llvm-project#55528]
The `perf` profiling integration in `lli` does not work with the ORC jit engine. There are no JitDump files created in ~/.debug/jit/. Hence, `perf report` does not know function addresses and boundaries of jitted code.

The following steps work for MCJIT (`--jit-kind=mcjit`), but not ORC. LLVM needs to be build with perf support: `LLVM_USE_PERF=ON`.

Simple C program `pi.c` to test profiling:

```c
#include <stdlib.h>
#include <stdio.h>

static double pi(size_t n){
        double res = 4.0;
        for(size_t i=0,v=3; i<n; ++i,v+=4){
                res -= 4.0 / (v);
                res += 4.0 / (v+2);
        }
        return res;
}

int main(int argc, char *argv[]){
        if(argc != 2){
                puts("requires n as input");
                return -1;
        }
        size_t n = atoi(argv[1]);
        double p = pi(n);
        printf("%lu: %f\n", n, p);
        return 0;
}
```

Convert to LLVM IR:

```console
$ clang -S -emit-llvm pi.c -o pi.ll
```

Profile execution:

```console
$ perf record -k 1 lli --jit-kind=mcjit pi.ll 100000000
$ perf inject -j -i perf.data -o perf.data.jitted
$ perf report -i perf.data.jitted
```

Without `--jit-kind=mcjit`, ORC is used by default which does not produce dump files. Therefore, `perf report` does not know about the jitted code.

**Comment by lhames** [comment]
This is a known issue.

Where ORC is using `RTDyldObjectLinkingLayer` the solution is just to register the JIT event listeners, similarly to how they're registered with the `ExecutionEngine` instance at: https://github.com/llvm/llvm-project/blob/5b00d13c00712de5161935d3d6b7af55cb8f3eb5/llvm/tools/lli/lli.cpp#L605

Where ORC is using `ObjectLinkingLayer` we don't have profiling support yet. We need volunteers to translate the existing JIT event listeners to `ObjectLinkingLayer::Plugins`.

**Comment by tetzank** [comment]
Thanks for the quick response. The registration was really just missing for ORC. With it, everything works fine.

```diff
diff --git a/llvm/tools/lli/lli.cpp b/llvm/tools/lli/lli.cpp
index f8c2c9172102..c99865e07862 100644
--- a/llvm/tools/lli/lli.cpp
+++ b/llvm/tools/lli/lli.cpp
@@ -944,9 +944,22 @@ int runOrcJIT(const char *ProgName) {
   auto J = ExitOnErr(Builder.create());
 
   auto *ObjLayer = &J->getObjLinkingLayer();
-  if (auto *RTDyldObjLayer = dyn_cast<orc::RTDyldObjectLinkingLayer>(ObjLayer))
+  if (auto *RTDyldObjLayer = dyn_cast<orc::RTDyldObjectLinkingLayer>(ObjLayer)) {
     RTDyldObjLayer->registerJITEventListener(
         *JITEventListener::createGDBRegistrationListener());
+#if LLVM_USE_OPROFILE
+    RTDyldObjLayer->registerJITEventListener(
+        *JITEventListener::createOProfileJITEventListener());
+#endif
+#if LLVM_USE_INTEL_JITEVENTS
+    RTDyldObjLayer->registerJITEventListener(
+        *JITEventListener::createIntelJITEventListener());
+#endif
+#if LLVM_USE_PERF
+    RTDyldObjLayer->registerJITEventListener(
+        *JITEventListener::createPerfJITEventListener());
+#endif
+  }
 
   if (PerModuleLazy)
     J->setPartitionFunction(orc::CompileOnDemandLayer::compileWholeModule);
```
Let's see if I can get this small patch submitted somehow.

**Comment by lhames** [comment]
@tetzank reviews.llvm.org is the place to submit LLVM patches. Please set me as a reviewer.

Profiling will break again when we switch the default JIT linker for ELF/x86-64 to JITLink. There's an option to force use of RuntimeDyld (`-jit-linker default | rtdyld | jitlink`), but that's a stop-gap solution at best. We'll really need to reimplement profiling support as `ObjectLinkingLayer::Plugins` to solve the issue properly.

**Comment by tetzank** [comment]
https://reviews.llvm.org/D126214
I hopefully got this right.

**Comment by lhames** [comment]
https://reviews.llvm.org/D126214 landed in b6ae90b86c42.

I'm going to close this issue. We can use https://github.com/llvm/llvm-project/issues/30869 to track the general profiling support task.
